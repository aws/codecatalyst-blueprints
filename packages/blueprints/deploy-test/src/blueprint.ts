import { Environment, EnvironmentDefinition, AccountConnection, Role } from '@caws-blueprint-component/caws-environments';
import { SourceRepository, SourceFile } from '@caws-blueprint-component/caws-source-repositories';
import { Blueprint as ParentBlueprint, Options as ParentOptions } from '@caws-blueprint/blueprints.blueprint';
import defaults from './defaults.json';

export interface Options extends ParentOptions {
  /**
   * This is some information about what type of environment and what in the world an environment is.
   * @displayName This is the Environment Title Area
   * @collapsed
   */
  testEnvironment: EnvironmentDefinition<{
    /**
     * blah blah blah some comments about the account that i'm deploying into
     * @displayName This account connection has an overriden name
     * @collapsed
     */
    testAccountConnection: AccountConnection<{
      /**
       * Blah blah some information about the role that I expect
       * e.g. here's a copy-pastable policy: [to a link]
       * @displayName This role has an overriden name
       */
      testRole: Role<['admin', 'lambda', 's3', 'cloudfront']>;
    }>;
  }>;
}
/**
 * This is the actual blueprint class.
 * 1. This MUST be the only 'class' exported, as 'Blueprint'
 * 2. This Blueprint should extend another ParentBlueprint
 */
export class Blueprint extends ParentBlueprint {
  constructor(options_: Options) {
    super(options_);
    console.log(defaults);
    // helpful typecheck for defaults
    const typeCheck: Options = {
      outdir: this.outdir,
      ...defaults,
    };
    const options = Object.assign(typeCheck, options_);

    options.testEnvironment.testAccountConnection;

    // add a repository
    const repo = new SourceRepository(this, { title: 'deployApiTestRepo' });

    // example showing add files to the repository
    new SourceFile(repo, 'testfile.txt', 'this is a line of text');
    new Environment(this, options.testEnvironment);
  }
}
