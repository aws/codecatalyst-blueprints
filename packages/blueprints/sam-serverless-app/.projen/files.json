{
  "files": [
    ".gitattributes",
    ".gitignore",
    ".npmignore",
    ".projen/deps.json",
    ".projen/files.json",
    ".projen/tasks.json",
    "LICENSE",
    "src/blueprint-snapshot-driver.spec.ts",
    "src/snapshot-infrastructure/infrastructure.ts",
    "tsconfig.dev.json",
    "tsconfig.json"
  ],
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \"npx projen\"."
}
