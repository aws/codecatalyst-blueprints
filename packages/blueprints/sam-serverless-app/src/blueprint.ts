import { Blueprint as ParentBlueprint, Options as ParentOptions } from '@caws-blueprint/blueprints.blueprint';
import defaults from './defaults.json';
import { generateReadmeContents } from './readmeContents'
import { Environment } from '@caws-blueprint-component/caws-environments';
import { SourceRepository } from '@caws-blueprint-component/caws-source-repositories';
import { SampleWorkspaces, Workspace } from '@caws-blueprint-component/caws-workspaces';
import {
  //generateWorkflow,
  WorkflowDefinition,
  Workflow,
  addGenericCloudFormationDeployAction,
  addGenericBranchTrigger,
  addGenericBuildAction,
  StageDefinition
} from '@caws-blueprint-component/caws-workflows';
import { SampleDir, SampleFile } from 'projen';
import * as cp from 'child_process';
import * as path from 'path';
import * as fs from 'fs';

import { RuntimeMapping } from './models';
import { runtimeMappings } from './runtimeMappings';

/**
 * This is the 'Options' interface. The 'Options' interface is interpreted by the wizard to dynamically generate a selection UI.
 * 1. It MUST be called 'Options' in order to be interpreted by the wizard
 * 2. This is how you control the fields that show up on a wizard selection panel. Keeping this small leads to a better user experience.
 * 3. You can use JSDOCs and annotations such as: '?', @advanced, @hidden, @display - textarea, etc. to control how the wizard displays certain fields.
 */
 export interface Options extends ParentOptions {
  /**
   * @displayName Runtime Language
   */
  runtime: 'Python 3' | 'Node.js 14' | 'Java 11 Maven' | 'Java 11 Gradle';

  /**
   * The name of the AWS CloudFormation stack generated for the blueprint. It must be unique for the AWS account it's being deployed to.
   * @displayName CloudFormation stack same
   * @validationRegex /^[a-zA-Z][a-zA-Z0-9-]{1,100}$/
   * @validationMessage Stack names must start with a letter, then contain alphanumeric characters and dashes(-) up to a total length of 128 characters
   * @defaultEntropy
   */
  cloudFormationStackName: string;

  /**
   * The configurations for your workflow
   */
  workflow: {

    /**
     * Configure default environment for this blueprint.
     */
    stages: StageDefinition[];

    /**
     * Enter the name of the S3 bucket to store build artifacts.
     * Must be an existing S3 bucket
     * @validationRegex /^(?!xn--)^(?!([0-9]{1,3}.){3}[0-9]{1,3}$)([a-z0-9][a-z0-9.-]{1,61}[a-z0-9])(?<!-s3alias)$/
     * @validationMessage SBucket names must only contain lowercase letters, numbers, and dashes.
     * See rules for bucket naming: https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucketnamingrules.html
     */
    s3BucketName: string;

    /**
     * Enter the role ARN to use when deploying your application through CloudFormation
     * @validationRegex /^arn:aws:iam::[0-9]{12}:role/[a-zA-Z0-9+=,.@_-]{1,64}$/
     * @validationMessage IAM role ARN must match pattern arn:aws:iam::<account ID>:role/<role name>.
     * Valid role names are up to 64 alphanumeric characters including the following symbols plus (+), equal (=), comma (,), period (.), at (@), underscore (_), and hyphen (-).
     */
     stackRoleArn: string;

    /**
     * Enter the role ARN to use when building your application
     * @validationRegex /^arn:aws:iam::[0-9]{12}:role/[a-zA-Z0-9+=,.@_-]{1,64}$/
     * @validationMessage IAM role ARN must match pattern arn:aws:iam::<account ID>:role/<role name>.
     * Valid role names are up to 64 alphanumeric characters including the following symbols plus (+), equal (=), comma (,), period (.), at (@), underscore (_), and hyphen (-).
     */
    buildRoleArn: string;
  }

  /**
   * @displayName Code Repository name
   * @collapsed
   */
  code: {
    /**
     * @displayName Code Repository name
     * @validationRegex /^[a-zA-Z0-9_.-]{1,100}$(?<!.git$)/
     * @validationMessage Must contain only alphanumeric characters, periods (.), underscores (_), dashes (-) and be up to 100 characters in length. Cannot end in .git or contain spaces
     */
    sourceRepositoryName: string;
  }


  /**
   * @displayName Lambda function name
   * @collapsed
   */
  lambda: {
    /**
     * Lambda function name must be unqiue to the AWS account it's being deployed to.
     * @displayName Lambda function name
     * @validationRegex /^[a-zA-Z0-9]{1,56}$/
     * @validationMessage Must contain only alphanumeric characters and be up to 56 characters in length
     */
    functionName: string;
  };
}

/**
 * This is the actual blueprint class.
 * 1. This MUST be the only 'class' exported, as 'Blueprint'
 * 2. This Blueprint should extend another ParentBlueprint
 */
 export class Blueprint extends ParentBlueprint {
   protected options: Options;
   protected readonly repository: SourceRepository;
    constructor(options_: Options) {
      super(options_);
      console.log(defaults);
      /**
       * This is a typecheck to ensure that the defaults passed in are of the correct type.
       * There are some cases where the typecheck will fail, but the defaults will still be valid, such when using enums.
       * you can override this ex. myEnum: defaults.myEnum as Options['myEnum'],
       */
      const typeCheck: Options = {
        outdir: this.outdir,
        ...defaults,
        runtime: defaults.runtime as Options['runtime'],
      };
      const options = Object.assign(typeCheck, options_);
      this.options = options;

      this.repository = new SourceRepository(this, {
        title: this.options.code.sourceRepositoryName || 'sam-lambda',
      });
      this.options.lambda = options.lambda;
  }

    override synth(): void {
      //MDE config workspace
      new Workspace(this, this.repository, SampleWorkspaces.default);
      //environments
      for (const stage of this.options.workflow.stages) {
        // if (stage.environment.title.length < 1){
        //   throw new Error('Invalid environment title length');
        // }
        stage.environment.title = `${stage.environment.title}`;
      }
      this.options.workflow.stages.forEach(stage => new Environment(this, stage.environment));

      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const runtimeOptions = runtimeMappings.get(this.options.runtime)!;
      const defaultReleaseBranch = 'main';
      const workflowName = 'build-and-release'
      //Workflow
      const workflowDefinition: WorkflowDefinition = {
        Name: workflowName,
        Triggers: [],
        Actions: {},
      }
      //todo: add region selection, when available
      const region = 'us-west-2'
      const artifactName = 'MyServerlessAppArtifact';
      this.addSamInstallScript();
      addGenericBranchTrigger(workflowDefinition, [defaultReleaseBranch]);
      addGenericBuildAction(this, workflowDefinition, this.options.workflow.buildRoleArn, [
        {Run: '. ./.aws/scripts/setup-sam.sh'},
        {Run: 'sam build'},
        {Run: `sam package --template-file ./.aws-sam/build/template.yaml --s3-bucket ${this.options.workflow.s3BucketName} --output-template-file output.yaml --region ${region}`},
      ], artifactName);
      //each stage should depend on the previous stage
      let previousStage = 'Build'
      for (const stage of this.options.workflow.stages) {
        // Append the environment title to the cloudformation stack name
        const cfnStackname = `${this.options.cloudFormationStackName}-${stage.environment.title}`
        //if (cfnStackname.length > 128) {
          //! Error message requires doc writer review
          //throw new Error ('Cloudformation stack name cannot be more than 128 characters')
        //}
        const cfnStage = {...stage, stackRoleArn: this.options.workflow.stackRoleArn}
        addGenericCloudFormationDeployAction(this, workflowDefinition, cfnStage, cfnStackname, region, previousStage, artifactName);
        previousStage = `Deploy_${stage.environment.title}`;
      }
      new Workflow(
        this,
        this.repository,
        workflowDefinition
      );
      this.createSamTemplate(runtimeOptions);

      const readmeContents = generateReadmeContents(runtimeOptions, defaultReleaseBranch, [this.options.lambda],
        this.options.workflow.stages, this.options.cloudFormationStackName, this.options.workflow.s3BucketName, workflowName);
      new SampleFile(this, path.join(this.repository.relativePath, 'README.md'),
      { contents: readmeContents });

      const toDeletePath = this.populateLambdaSourceCode(runtimeOptions);
      super.synth();
      // verify synth generating source code for each lambda
      for (const lambdaName of [this.options.lambda?.functionName]) {
        if (!fs.existsSync(path.join(this.repository.path, lambdaName || ''))) {
          //! Error message requires doc writer review
          throw new Error('Unable to synthesize source code for lambda function');
        }
      }
      cp.execSync(`rm -rf ${toDeletePath}`);
   }

   /**
    * Populates source code for lambda functions.
    * Source code is checked out from sam templates
    */
   protected populateLambdaSourceCode(runtimeOptions: RuntimeMapping): string{
      const sourceDir = path.join('/tmp/sam-lambdas', runtimeOptions?.cacheDir);
      const runtime = runtimeOptions?.runtime;
      const gitSrcPath = runtimeOptions?.gitSrcPath;

      cp.execSync(`svn checkout https://github.com/aws/aws-sam-cli-app-templates/trunk/${runtime}/${gitSrcPath}/\{\{cookiecutter.project_name\}\} ${sourceDir}; \
      rm -rf ${sourceDir}/.svn ${sourceDir}/.gitignore ${sourceDir}/README.md ${sourceDir}/template.yaml`);

      for (const lambda of [this.options.lambda?.functionName]) {
        const newLambdaPath = path.join(this.repository.relativePath, lambda || '');
        new SampleDir(this, newLambdaPath, {sourceDir});
      }
      return sourceDir;
  }

  protected addSamInstallScript() {
    new SampleFile(this, path.join(this.repository.relativePath, '.aws', 'scripts', 'setup-sam.sh'), {
      contents: `#!/usr/bin/env bash
echo "Setting up sam"

yum install unzip -y

curl -LO https://github.com/aws/aws-sam-cli/releases/latest/download/aws-sam-cli-linux-x86_64.zip
unzip -qq aws-sam-cli-linux-x86_64.zip -d sam-installation-directory

./sam-installation-directory/install; export AWS_DEFAULT_REGION=us-west-2
`,
    });
  }

   /**
    * Generate Sam Template
    */
   protected createSamTemplate(runtimeOptions: RuntimeMapping): void {
    const header =
`Transform: AWS::Serverless-2016-10-31
Description: lambdas
Globals:
  Function:
    Timeout: 20\n`;
    let resources = 'Resources:';
    let outputs = 'Outputs:';
    for (const lambda of [this.options.lambda?.functionName]) {
       resources += `
  ${lambda}Function:
    Type: AWS::Serverless::Function
    Properties:
      CodeUri: ${lambda}/${runtimeOptions?.codeUri}
      Runtime: ${runtimeOptions?.runtime}
      Handler: ${runtimeOptions?.handler}
      Description: ${lambda}
      Events:
          ${lambda}:
             Type: Api # More info about API Event Source: https://github.com/awslabs/serverless-application-model/blob/master/versions/2016-10-31.md#api
             Properties:
                Path: /${lambda}
                Method: get`
      //Append additional template properties
      resources += runtimeOptions?.templateProps;

       outputs += `
# ServerlessRestApi is an implicit API created out of Events key under Serverless::Function
# Find out more about other implicit resources you can reference within SAM
# https://github.com/awslabs/serverless-application-model/blob/master/docs/internals/generated_resources.rst#api
  ${lambda}Api:
    Description: "API Gateway endpoint URL for Prod stage for Hello World function"
    Value: !Sub "https://\${ServerlessRestApi}.execute-api.\${AWS::Region}.amazonaws.com/Prod/${lambda}/\"
  ${lambda}Function:
    Description: "Hello World Lambda Function ARN"
    Value: !GetAtt ${lambda}Function.Arn
  ${lambda}FunctionIamRole:
    Description: "Implicit IAM Role created for Hello World function"
    Value: !GetAtt ${lambda}FunctionRole.Arn`
    }

    const destinationPath = path.join(this.repository.relativePath, 'template.yaml');
    const template = header + resources + '\n' + outputs;
    new SampleFile(this, destinationPath, { contents: template });
 }

}



