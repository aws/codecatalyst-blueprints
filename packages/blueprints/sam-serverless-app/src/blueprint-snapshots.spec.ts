// ~~ This file is owned by caws-blueprints and is generated by projen. To modify, edit .projenrc.js and run "npx projen".
import * as fs from 'fs';

import { Blueprint, Options } from './blueprint';
import { cleanUpOutdir, prepareNewOutdir } from './testSnapshotInfra/outdir';
import { allTestConfigs } from './testSnapshotInfra/testConfigs';
import { getAllBlueprintSnapshottedFilenames } from './testSnapshotInfra/files';

describe('Blueprint snapshots for test configurations', () => {
  allTestConfigs().forEach(testConfig => {
    describe(`${testConfig.name} configuration`, () => {
      let blueprint;

      beforeAll(() => {
        const outdir = prepareNewOutdir();
        const options: Options = {
          ...testConfig.config,
          outdir,
        };
        blueprint = new Blueprint(options);
        blueprint.synth();
      });

      afterAll(() => {
        cleanUpOutdir(blueprint.outdir);
      });

      it('matches snapshots', async () => {
        for await (const snappedFile of getAllBlueprintSnapshottedFilenames(blueprint.outdir)) {
          await expect(fs.readFileSync(snappedFile.absPath, { encoding: 'utf-8' })).toMatchSnapshot(snappedFile.relPath);
        }
      });
    });
  });
});
