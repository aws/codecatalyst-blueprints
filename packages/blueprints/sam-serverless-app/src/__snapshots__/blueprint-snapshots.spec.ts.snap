// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blueprint snapshots for test configurations default-config configuration matches snapshots: src/ServerlessAppRepo/.codecatalyst/scripts/run-tests.sh 1`] = `
"#!/usr/bin/env bash

WORKING_DIR=SamFirstEndpoint/hello-world/
npm install --prefix $WORKING_DIR
npm --prefix $WORKING_DIR run coverage
"
`;

exports[`Blueprint snapshots for test configurations default-config configuration matches snapshots: src/ServerlessAppRepo/.codecatalyst/workflows/build-and-release.yaml 1`] = `
"
Name: build-and-release
SchemaVersion: "1.0"
Triggers:
  - Type: PUSH
    Branches:
      - main
Compute:
  Type: Lambda
  Fleet: Linux.x86-64.Large
Actions:
  build_for_default_environment:
    Identifier: aws/build-beta@v1
    Inputs:
      Sources:
        - WorkflowSource
    Outputs:
      AutoDiscoverReports:
        Enabled: true
        ReportNamePrefix: rpt
      Artifacts:
        - Name: build_result
          Files:
            - "**/*"
    Configuration:
      Steps:
        - Run: . ./.codecatalyst/scripts/run-tests.sh
        - Run: sam build --template-file template.yaml
        - Run: cd .aws-sam/build/
        - Run: sam package --output-template-file packaged.yaml --resolve-s3 --template-file template.yaml --region us-west-2
    Environment:
      Name: default_environment
      Connections:
        - Name: " "
          Role: " "
  deploy_to_default_environment:
    Identifier: aws/cfn-deploy-gamma@v1
    Inputs:
      Artifacts:
        - build_result
    Environment:
      Name: default_environment
      Connections:
        - Name: " "
          Role: " "
    Configuration:
      Parameters:
        region: us-west-2
        name: sam-stack-
        template: .aws-sam/build/packaged.yaml
        no-fail-on-empty-changeset: "1"
        capabilities: CAPABILITY_IAM,CAPABILITY_NAMED_IAM,CAPABILITY_AUTO_EXPAND
"
`;

exports[`Blueprint snapshots for test configurations default-config configuration matches snapshots: src/ServerlessAppRepo/README.md 1`] = `
"
## This Project:

This project is an AWS Serverless Application Model (SAM) Project generated using the AWS Blueprints System. For more information on blueprints, see the _Working with blueprints in Quokka_ section in the **Quokka User Guide**. This project contains the source code, supporting files, and Quokka.Codes resources for a SAM application that is deployed using Quokka.Codes workflows.

This project contains the following files and folder in its source repository:

  - SamFirstEndpoint - Source code and supporting files for the Lambda function of the application, containing:

    - hello-world - Code for the Lambda function of the application

    - events - Invocation events that you can use to invoke the function

    - hello-world/tests - Tests for the Lambda function's code

  - .codecatalyst/workflows/build-and-release.yaml - The template that defines the project's workflow

  - template.yaml - The template that defines the application's AWS resources, including Lambda Functions, API Gateways, and IAM roles

  - .mde.devfile.yaml - A DevFile that defines developer workspaces or cloud-native development environments.

This project has created the following Quokka.Codes Resources:

- Source repository - A git repository to store, version, and manage project assets.

  For more information on source repositories, see the _Source Repositories in Quokka_ section in the **Quokka User Guide**

- Workflow - An automated procedure that defines how to build, test, and deploy the serverless application.

  For more information on workflows, see the _Build, test, and deploy with workflows in Quokka_ section of the **Quokka User Guide**

- Environment(s) - An abstraction of infrastructure resources for deploying applications. Environments can be used to organize deployment actions into a production or non-production environment

  For more information on environments, see the _Organizing deployments using environments_ section in the **Quokka User Guide**

- Workspace - A cloud native development environment. Workspace must be manually created with the generated DevFile using the create workspace operation on Quokka.Codes.

  For more information on the create workspace operation and workspaces, see the _Working with workspaces in Quokka_ section in the **Quokka User Guide**

This project will deploy the following AWS Resources after being successfuly created, the deployment status can be viewed in the project's workflow:

  - Lambda Function(s) - A resource to invoke your code on a high-availability compute infrastructure without provisioning or managing servers. For more information on AWS Lambda, see the [AWS Lambda Developer Guide](https://docs.aws.amazon.com/lambda/latest/dg/welcome.html)

  - API Gateway - A resource for creating, publishing, maintaining, monitoring, and securing REST, HTTP, and WebSocket APIs at any scale. For more information on API Gateway, see the [AWS API Gateway Developer Guide](https://docs.aws.amazon.com/apigateway/latest/developerguide/welcome.html)

  - IAM Role(s) - A resource for securely controlled access to AWS resource, such as the lambda function(s). For more information on IAM, see the [AWS IAM User Guide](https://docs.aws.amazon.com/IAM/latest/UserGuide/introduction.html)

## Serverless Application

A serverless application is a combination of Lambda functions, event sources, and other resources that work together to perform tasks. Note that a serverless application is more than just a Lambda functionâ€”it can include additional resources such as APIs, databases, and event source mappings.
For more information on serverless applications, see the [AWS Serverless Application Model (SAM) Developer Guide](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/what-is-sam.html)

## Using the SAM CLI to build and test locally

You can use the Serverless Application Model Command Line Interface (SAM CLI) to build and test your application locally. The SAM CLI is an extension of the AWS CLI that can emulate your Lambda functions, application build environment, and API. It uses Docker to run your functions in an Amazon Linux environment that matches Lambda. It can also emulate your application's build environment and API.
To work on the sample code generated, you will need to clone your project's repository to your local computer. If you haven't, do that first. You can find instructions in the _Clone a source repository_ section in the **Quokka User Guide**.

To use the SAM CLI, you need the following tools into your workspace.

  * Install [AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html)
  * Install [SAM CLI](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-cli-install.html)
  * Install [Docker community edition](https://hub.docker.com/search/?type=edition&offering=community)
  * Install [Python 3](https://www.python.org/downloads/)
 * Install [Node.js 14 and npm](https://nodejs.org/en/download/releases/)

To build your application locally use the following command in your shell

\`\`\`
   sam build
\`\`\`

  The SAM CLI installs dependencies defined in the hello-world//package.json file of each lambda functions, creates a deployment package, and saves it in the .aws-sam/build folder.
  For more information on sam build, see the [Sam Build Command Reference Guide](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-cli-command-reference-sam-build.html).
  Test a single function by invoking it directly with a test event. An event is a JSON document that represents the input that the function receives from the event source. Test events are included in the \`events\` folder in each function's folder in this project.

\`\`\`
  sam local invoke <functionName> --event <functionName>/events/event.json
\`\`\`

For more information on sam local invoke, see the [Sam Invoke Command Reference Guide](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-cli-command-reference-sam-local-invoke.html).
The SAM CLI can also emulate your applications API. Use the sam local start-api to run the API locally (The default port is 3000).

\`\`\`
  sam local start-api
  curl http://localhost:3000/
\`\`\`

For more information on sam local start-api, see the [Sam Local Invoke Start-Api Command Reference Guide](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-cli-command-reference-sam-local-start-api.html).
The SAM CLI reads the application template to determine the API's routes and the functions that they invoke. The \`Events\` property on each function's definition includes the route and method for each path.

\`\`\`yaml
      Events:
        HelloWorld:
          Type: Api
          Properties:
            Path: /hello
            Method: get
\`\`\`


## Tests
Tests are defined in the \`hello-world/tests\` folder in this project. Use NPM to install the [Mocha test framework](https://mochajs.org/) and run unit tests.
\`\`\`
$ cd hello-world
$ npm install
$ npm run test
\`\`\`


## Add a resource to your serverless application
The application template uses SAM to define application resources. AWS SAM is an extension of AWS CloudFormation with a simpler syntax for configuring common serverless application resources such as functions, triggers, and APIs. For resources not included in the [SAM specification](https://github.com/awslabs/serverless-application-model/blob/master/versions/2016-10-31.md), you can use standard [AWS CloudFormation](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-template-resource-type-ref.html) resource types.

## Deploying your serverless application
The application is deployed through Quokka.Codes using the workflow defined in \`.codecatalyst/workflows/build-and-release.yaml\`. The workflow is triggered by pushes to the \`main\` of the source repository. Triggers can be code pushes to a source repository branch or a pull request being created, merged, closed, or revised. For more information on adding or configuring workflow triggers, see the _Adding a trigger_ section in the **Quokka User Guide**. The workflow builds your application, stores the build artifacts in a generated s3 bucket, and deploys your application to your project environments in the following order:

  - \`default_environment\` using the cloudformation stack \`sam-stack--default_environment\`


For more information on deploying using workflows and organizing deployments by environment, see the _Deploying using workflows_ section in the **Quokka User Guide**.

## Additional Resources
See the Quokka User Guide for additional information on using the features and resources of Quokka.Codes
"
`;

exports[`Blueprint snapshots for test configurations default-config configuration matches snapshots: src/ServerlessAppRepo/SamFirstEndpoint/events/event.json 1`] = `
"{
  "body": "{\\"message\\": \\"hello world\\"}",
  "resource": "/{proxy+}",
  "path": "/path/to/resource",
  "httpMethod": "POST",
  "isBase64Encoded": false,
  "queryStringParameters": {
    "foo": "bar"
  },
  "pathParameters": {
    "proxy": "/path/to/resource"
  },
  "stageVariables": {
    "baz": "qux"
  },
  "headers": {
    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8",
    "Accept-Encoding": "gzip, deflate, sdch",
    "Accept-Language": "en-US,en;q=0.8",
    "Cache-Control": "max-age=0",
    "CloudFront-Forwarded-Proto": "https",
    "CloudFront-Is-Desktop-Viewer": "true",
    "CloudFront-Is-Mobile-Viewer": "false",
    "CloudFront-Is-SmartTV-Viewer": "false",
    "CloudFront-Is-Tablet-Viewer": "false",
    "CloudFront-Viewer-Country": "US",
    "Host": "1234567890.execute-api.us-east-1.amazonaws.com",
    "Upgrade-Insecure-Requests": "1",
    "User-Agent": "Custom User Agent String",
    "Via": "1.1 08f323deadbeefa7af34d5feb414ce27.cloudfront.net (CloudFront)",
    "X-Amz-Cf-Id": "cDehVQoZnx43VYQb9j2-nvCh-9z396Uhbp027Y2JvkCPNLmGJHqlaA==",
    "X-Forwarded-For": "127.0.0.1, 127.0.0.2",
    "X-Forwarded-Port": "443",
    "X-Forwarded-Proto": "https"
  },
  "requestContext": {
    "accountId": "123456789012",
    "resourceId": "123456",
    "stage": "prod",
    "requestId": "c6af9ac6-7b61-11e6-9a41-93e8deadbeef",
    "requestTime": "09/Apr/2015:12:34:56 +0000",
    "requestTimeEpoch": 1428582896000,
    "identity": {
      "cognitoIdentityPoolId": null,
      "accountId": null,
      "cognitoIdentityId": null,
      "caller": null,
      "accessKey": null,
      "sourceIp": "127.0.0.1",
      "cognitoAuthenticationType": null,
      "cognitoAuthenticationProvider": null,
      "userArn": null,
      "userAgent": "Custom User Agent String",
      "user": null
    },
    "path": "/prod/path/to/resource",
    "resourcePath": "/{proxy+}",
    "httpMethod": "POST",
    "apiId": "1234567890",
    "protocol": "HTTP/1.1"
  }
}
"
`;

exports[`Blueprint snapshots for test configurations default-config configuration matches snapshots: src/ServerlessAppRepo/SamFirstEndpoint/hello-world/app.js 1`] = `
"// const axios = require('axios')
// const url = 'http://checkip.amazonaws.com/';
let response;

/**
 *
 * Event doc: https://docs.aws.amazon.com/apigateway/latest/developerguide/set-up-lambda-proxy-integrations.html#api-gateway-simple-proxy-for-lambda-input-format
 * @param {Object} event - API Gateway Lambda Proxy Input Format
 *
 * Context doc: https://docs.aws.amazon.com/lambda/latest/dg/nodejs-prog-model-context.html 
 * @param {Object} context
 *
 * Return doc: https://docs.aws.amazon.com/apigateway/latest/developerguide/set-up-lambda-proxy-integrations.html
 * @returns {Object} object - API Gateway Lambda Proxy Output Format
 * 
 */
exports.lambdaHandler = async (event, context) => {
    try {
        // const ret = await axios(url);
        response = {
            'statusCode': 200,
            'body': JSON.stringify({
                message: 'hello world',
                // location: ret.data.trim()
            })
        }
    } catch (err) {
        console.log(err);
        return err;
    }

    return response
};
"
`;

exports[`Blueprint snapshots for test configurations default-config configuration matches snapshots: src/ServerlessAppRepo/SamFirstEndpoint/hello-world/package.json 1`] = `
"{
  "name": "hello_world",
  "version": "1.0.0",
  "description": "hello world sample for NodeJS",
  "main": "app.js",
  "repository": "https://github.com/awslabs/aws-sam-cli/tree/develop/samcli/local/init/templates/cookiecutter-aws-sam-hello-nodejs",
  "author": "SAM CLI",
  "license": "MIT",
  "dependencies": {
    "axios": "^0.21.1"
  },
  "scripts": {
    "test": "mocha --reporter mocha-junit-reporter tests/unit/",
    "coverage": "nyc --reporter clover --reporter text npm test"
  },
  "devDependencies": {
    "chai": "^4.2.0",
    "mocha": "^9.1.4",
    "mocha-junit-reporter": "^2.0.2",
    "nyc": "^15.1.0"
  }
}
"
`;

exports[`Blueprint snapshots for test configurations default-config configuration matches snapshots: src/ServerlessAppRepo/SamFirstEndpoint/hello-world/tests/unit/test-handler.js 1`] = `
"'use strict';

const app = require('../../app.js');
const chai = require('chai');
const expect = chai.expect;
var event, context;

describe('Tests index', function () {
    it('verifies successful response', async () => {
        const result = await app.lambdaHandler(event, context)

        expect(result).to.be.an('object');
        expect(result.statusCode).to.equal(200);
        expect(result.body).to.be.an('string');

        let response = JSON.parse(result.body);

        expect(response).to.be.an('object');
        expect(response.message).to.be.equal("hello world");
        // expect(response.location).to.be.an("string");
    });
});
"
`;

exports[`Blueprint snapshots for test configurations default-config configuration matches snapshots: src/ServerlessAppRepo/devfile.yaml 1`] = `
"
schemaVersion: 2.0.0
metadata:
  name: aws-universal
  version: 1.0.1
  displayName: AWS Universal
  description: Stack with AWS Universal Tooling
  tags:
    - aws
    - a12
  projectType: aws
components:
  - name: aws-runtime
    container:
      image: public.ecr.aws/aws-mde/universal-image:latest
      mountSources: true
      volumeMounts:
        - name: docker-store
          path: /var/lib/docker
  - name: docker-store
    volume:
      size: 16Gi
"
`;

exports[`Blueprint snapshots for test configurations default-config configuration matches snapshots: src/ServerlessAppRepo/template.yaml 1`] = `
"Transform: AWS::Serverless-2016-10-31
Description: lambdas
Globals:
  Function:
    Timeout: 20
Resources:
  SamFirstEndpointFunction:
    Type: AWS::Serverless::Function
    Properties:
      CodeUri: SamFirstEndpoint/hello-world/
      Runtime: nodejs14.x
      Handler: app.lambdaHandler
      Description: SamFirstEndpoint
      Events:
          SamFirstEndpoint:
             Type: Api # More info about API Event Source: https://github.com/awslabs/serverless-application-model/blob/master/versions/2016-10-31.md#api
             Properties:
                Path: /SamFirstEndpoint
                Method: get
Outputs:
# ServerlessRestApi is an implicit API created out of Events key under Serverless::Function
# Find out more about other implicit resources you can reference within SAM
# https://github.com/awslabs/serverless-application-model/blob/master/docs/internals/generated_resources.rst#api
  SamFirstEndpointApi:
    Description: "API Gateway endpoint URL for Prod stage for Hello World function"
    Value: !Sub "https://\${ServerlessRestApi}.execute-api.\${AWS::Region}.amazonaws.com/Prod/SamFirstEndpoint/"
  SamFirstEndpointFunction:
    Description: "Hello World Lambda Function ARN"
    Value: !GetAtt SamFirstEndpointFunction.Arn
  SamFirstEndpointFunctionIamRole:
    Description: "Implicit IAM Role created for Hello World function"
    Value: !GetAtt SamFirstEndpointFunctionRole.Arn"
`;
